qiskit/circuit/library/arithmetic/multipliers/rg_qft_multiplier.py

```diff
@@ -0,0 +1,82 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Compute the product of two qubit registers using QFT."""
+
+from typing import Optional
+import numpy as np
+
+from qiskit.circuit import QuantumRegister
+from qiskit.circuit.library.standard_gates import PhaseGate
+from qiskit.circuit.library.basis_change import QFT
+
+from .multiplier import Multiplier
+
+
+class RGQFTMultiplier(Multiplier):
+    r"""A QFT multiplication circuit to store product of two input registers out-of-place.
+
+    Multiplication in this circuit is implemented using the procedure of Fig. 3 in [1], where
+    weighted sum rotations are implemented as given in Fig. 5 in [1]. QFT is used on the output
+    register and is followed by rotations controlled by input registers. The rotations
+    transform the state into the product of two input registers in QFT base, which is
+    reverted from QFT base using inverse QFT.
```

Comment:
Could you also add a circuit diagram here? Or is the circuit too large to be shown?

Reply:
Non-modular circuit was too large even for 2 qubits 😄 Modular one for 2-qubit input registers and 2-qubit output register barely fits:
```
          a_0: ────────────────────────────────────────■───────■──────■──────■────────────────
                                                       │       │      │      │
          a_1: ─────────■───────■───────■───────■──────┼───────┼──────┼──────┼────────────────
                        │       │       │       │      │       │      │      │
          b_0: ─────────┼───────┼───────■───────■──────┼───────┼──────■──────■────────────────
                        │       │       │       │      │       │      │      │
          b_1: ─────────■───────■───────┼───────┼──────■───────■──────┼──────┼────────────────
               ┌──────┐ │P(4π)  │       │P(2π)  │      │P(2π)  │      │P(π)  │       ┌───────┐
        out_0: ┤0     ├─■───────┼───────■───────┼──────■───────┼──────■──────┼───────┤0      ├
               │  qft │         │P(2π)          │P(π)          │P(π)         │P(π/2) │  iqft │
        out_1: ┤1     ├─────────■───────────────■──────────────■─────────────■───────┤1      ├
               └──────┘                                                              └───────┘
```

I imagine that a modular circuit of such a small size would never be used in reality, is it still fine to include that as an example then?

Reply:
Yeah I think you should include it, just to give developers and users and idea of what components the circuit has 👍🏻

Reply:
For the modular one we can use jupyter to at least show the circuit diagram in the online documentation, like so

```
    .. jupyter-execute::
        :hide-code:

        from qiskit.circuit.library import TheMultiplier
        import qiskit.tools.jupyter
        circuit = TheMultiplier(2)
        %circuit_library_info circuit
```

Reply:
I have added a static circuit diagram, since the modular one for 2 input and output qubit registers was within 100 characters 👍